File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/property/TextAppearance.java
/* Copyright Airship and Contributors */

package com.urbanairship.android.layout.property;

import com.urbanairship.json.JsonException;
import com.urbanairship.json.JsonList;
import com.urbanairship.json.JsonMap;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.Dimension;
import androidx.annotation.NonNull;

public class TextAppearance {
    @NonNull
    private final Color color;
    @Dimension(unit = Dimension.DP)
    private final int fontSize;
    @NonNull
    private final TextAlignment alignment;
    @NonNull
    private final List<TextStyle> textStyles;
    @NonNull
    private final List<String> fontFamilies;

    public TextAppearance(
        @NonNull Color color,
        int fontSize,
        @NonNull TextAlignment alignment,
        @NonNull List<TextStyle> textStyles,
        @NonNull List<String> fontFamilies
    ) {
        this.color = color;
        this.fontSize = fontSize;
        this.alignment = alignment;
        this.textStyles = textStyles;
        this.fontFamilies = fontFamilies;
    }

    @NonNull
    public static TextAppearance fromJson(@NonNull JsonMap json) throws JsonException {
        int fontSize = json.opt("font_size").getInt(14);
        Color color = Color.fromJsonField(json, "color");
        if (color == null) {
            throw new JsonException("Failed to parse text appearance. 'color' may not be null!");
        }
        String alignmentString = json.opt("alignment").optString();
        JsonList textStylesJson = json.opt("styles").optList();
        JsonList fontFamiliesJson = json.opt("font_families").optList();
        TextAlignment alignment = alignmentString.isEmpty()
            ? TextAlignment.CENTER
            : TextAlignment.from(alignmentString);

        List<TextStyle> textStyles = new ArrayList<>();
        for (int i = 0; i < textStylesJson.size(); i++) {
            String styleString = textStylesJson.get(i).optString();
            TextStyle style = TextStyle.from(styleString);
            textStyles.add(style);
        }

        List<String> fontFamilies = new ArrayList<>();
        for (int i = 0; i < fontFamiliesJson.size(); i++) {
            String fontFamily = fontFamiliesJson.get(i).optString();
            fontFamilies.add(fontFamily);
        }

        return new TextAppearance(color, fontSize, alignment, textStyles, fontFamilies);
    }

    @NonNull
    public Color getColor() {
        return color;
    }

    @Dimension(unit = Dimension.DP)
    public int getFontSize() {
        return fontSize;
    }

    @NonNull
    public TextAlignment getAlignment() {
        return alignment;
    }

    @NonNull
    public List<TextStyle> getTextStyles() {
        return textStyles;
    }

    @NonNull
    public List<String> getFontFamilies() {
        return fontFamilies;
    }
}

File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/property/TextAppearance.kt
/* Copyright Airship and Contributors */
package com.urbanairship.android.layout.property

import androidx.annotation.Dimension
import androidx.annotation.IntRange
import com.urbanairship.json.JsonException
import com.urbanairship.json.JsonValue
import com.urbanairship.json.optionalField
import com.urbanairship.json.optionalList

public open class TextAppearance {

    public val color: Color

    @JvmField
    @Dimension(unit = Dimension.Companion.DP)
    public val fontSize: Int

    @JvmField
    public val alignment: TextAlignment

    @JvmField
    public val textStyles: List<TextStyle>

    @JvmField @IntRange(from = 0, to = 900)
    public val fontWeight: Int

    @JvmField
    public val fontFamilies: List<String>

    protected constructor(textAppearance: TextAppearance) {
        this.color = textAppearance.color
        this.fontSize = textAppearance.fontSize
        this.alignment = textAppearance.alignment
        this.textStyles = textAppearance.textStyles
        this.fontWeight = textAppearance.fontWeight
        this.fontFamilies = textAppearance.fontFamilies
    }

    public constructor(
        color: Color,
        fontSize: Int,
        alignment: TextAlignment,
        textStyles: List<TextStyle>,
        fontWeight: Int,
        fontFamilies: List<String>
    ) {
        this.color = color
        this.fontSize = fontSize
        this.alignment = alignment
        this.textStyles = textStyles
        this.fontWeight = fontWeight
        this.fontFamilies = fontFamilies
    }

    public companion object {
        private const val KEY_FONT_SIZE = "font_size"
        private const val KEY_COLOR = "color"
        private const val KEY_ALIGNMENT = "alignment"
        private const val KEY_STYLES = "styles"
        private const val KEY_FONT_WEIGHT = "font_weight"
        private const val KEY_FONT_FAMILIES = "font_families"

        @JvmStatic
        @Throws(JsonException::class)
        public fun fromJson(json: JsonValue): TextAppearance {
            val content = json.requireMap()

            return TextAppearance(
                color = Color.fromJson(content.require(KEY_COLOR)),
                fontSize = content.optionalField(KEY_FONT_SIZE) ?: 14,
                alignment = content[KEY_ALIGNMENT]?.let(TextAlignment::from) ?: TextAlignment.CENTER,
                textStyles = content.optionalList(KEY_STYLES)?.map(TextStyle::from) ?: emptyList(),
                fontWeight = content.optionalField(KEY_FONT_WEIGHT) ?: 0,
                fontFamilies = content.optionalList(KEY_FONT_FAMILIES)?.mapNotNull { it.string } ?: emptyList()
            )
        }
    }
}
